src/Map/Map.tsx

```diff
@@ -0,0 +1,166 @@
+
+  const long = props.startPosition?.long || -3.5
+  const lat = props.startPosition?.lat || 55
+  const zoom = props.startPosition?.zoom || 5
+
+  const style = props.style || 'mapbox://styles/mapbox/light-v11'
+  const source = props.sourceJson || ''
+  const cluster = props.cluster || false
+
+  const clusterMaxZoom = props.clusterOptions?.clusterMaxZoom || 14
+  const clusterAreaRadius = props.clusterOptions?.clusterAreaRadius || 50
+  const clusterColor = props.clusterOptions?.clusterColor || colors.black
+  const clusterRadius = props.clusterOptions?.clusterRadius || 20
+  const countFont = props.clusterOptions?.countFont || ['Open Sans Regular']
+  const countFontSize = props.clusterOptions?.countFontSize || 14
+  const countFontColor = props.clusterOptions?.countFontColor || colors.white
+  const pointColor = props.clusterOptions?.pointColor || colors.black
+  const pointRadius = props.clusterOptions?.pointRadius || 6
+
+  useEffect(() => {
+    if (map.current) return // initialize map only once
+    map.current = new mapboxgl.Map({
+      container: mapContainer.current!,
+      style: style,
+      center: [long, lat],
+      zoom: zoom,
+      attributionControl: false,
+    })
+  })
+
+  useEffect(() => {
+    if (!map.current) return // wait for map to initialize
+    map.current.on('load', () => {
+      map.current?.addSource('source', {
+        type: 'geojson',
+        data: source,
+        cluster: cluster,
+        clusterMaxZoom: clusterMaxZoom, // Max zoom to cluster points on
+        clusterRadius: clusterAreaRadius, // Radius of each cluster when clustering points (defaults to 50)
+      })
+
+      map.current?.addLayer({
+        id: 'clusters',
+        type: 'circle',
+        source: 'source',
+        filter: ['has', 'point_count'],
+        paint: {
+          'circle-color': clusterColor,
+          'circle-radius': clusterRadius,
+        },
+      })
+
+      map.current?.addLayer({
+        id: 'cluster-count',
+        type: 'symbol',
+        source: 'source',
+        filter: ['has', 'point_count'],
+        layout: {
+          'text-field': ['get', 'point_count_abbreviated'],
+          'text-font': countFont,
+          'text-size': countFontSize,
+        },
+        paint: {
+          'text-color': countFontColor,
+        },
+      })
+
+      map.current?.addLayer({
+        id: 'unclustered-point',
+        type: 'circle',
+        source: 'source',
+        filter: ['!', ['has', 'point_count']],
+        paint: {
+          'circle-color': pointColor,
+          'circle-radius': pointRadius,
+        },
+      })
+
+      // inspect a cluster on click
+      map.current?.on('click', 'clusters', (e) => {
+        const features = map.current!.queryRenderedFeatures(e.point, {
+          layers: ['clusters'],
+        })
+
+        const clusterId = features[0]?.properties?.cluster_id
+        const source: mapboxgl.GeoJSONSource = map.current!.getSource(
```

Comment:
nice, not used this operator `key1!.key2`

Reply:
what does this do?